pkg/agent/proxy/route.go

```diff
@@ -0,0 +1,301 @@
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// NodeReconciler watch node and update route table
+type NodeReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+
+	StopChan    <-chan struct{}
+	updateMutex sync.Mutex
+}
+
+func GetNodeInternalIP(node corev1.Node) net.IP {
+	for _, item := range node.Status.Addresses {
+		if item.Type == corev1.NodeInternalIP {
+			return net.ParseIP(item.Address)
+		}
+	}
+	return nil
+}
+
+func GetRouteByDst(dst *net.IPNet) []netlink.Route {
+	var ret []netlink.Route
+	// List all route item in current node
+	routeList, err := netlink.RouteList(nil, unix.AF_INET)
+	if err != nil {
+		klog.Errorf("List route table error, err:%s", err)
+		return ret
+	}
+	for _, item := range routeList {
+		if item.Dst != nil && item.Dst.String() == dst.String() {
+			ret = append(ret, item)
+		}
+	}
+	return ret
+}
+func RouteEqual(r1, r2 netlink.Route) bool {
+	return ((r1.Dst == nil && r2.Dst == nil) ||
+		(r1.Dst != nil && r2.Dst != nil && r1.Dst.String() == r2.Dst.String())) &&
+		r1.Src.Equal(r2.Src) &&
+		r1.Gw.Equal(r2.Gw)
+}
+
+func UpdateRoute(nodeList corev1.NodeList, thisNode corev1.Node) {
+	var oldRoute []netlink.Route
+	var newRoute []netlink.Route
+	var err error
+
+	for _, item := range nodeList.Items {
+		// ignore current node
+		if item.Name == thisNode.Name {
+			continue
+		}
+		gw := GetNodeInternalIP(item)
+		if gw == nil {
+			klog.Errorf("Fail to get node internal IP in node: %s", item.Name)
+			continue
+		}
+		// multi-podCIDRs will create multi-routeItem
+		for _, podCIDR := range item.Spec.PodCIDRs {
+			dst, err := netlink.ParseIPNet(podCIDR)
+			if err != nil {
+				klog.Errorf("Parse podCIDR %s failed, err: %s", podCIDR, err)
+			}
+			tempRoute := GetRouteByDst(dst)
+			if len(tempRoute) != 0 {
+				oldRoute = append(oldRoute, tempRoute...)
+			}
+			newRoute = append(newRoute, netlink.Route{
+				Dst:   dst,
+				Gw:    gw,
+				Table: 254,
+			})
+		}
+	}
+
+	var delRoute []netlink.Route
+	var addRoute []netlink.Route
+	// calculate route to delete
+	for _, oldItem := range oldRoute {
+		exist := false
+		for _, newItem := range newRoute {
+			if RouteEqual(oldItem, newItem) {
+				exist = true
+				break
+			}
+		}
+		if !exist {
+			delRoute = append(delRoute, oldItem)
+		}
+	}
+	// calculate route to add
+	for _, newItem := range newRoute {
+		exist := false
+		for _, oldItem := range oldRoute {
+			if RouteEqual(oldItem, newItem) {
+				exist = true
+				break
+			}
+		}
+		if !exist {
+			addRoute = append(addRoute, newItem)
+		}
+	}
+
+	// add & del route
+	for i := range delRoute {
+		if err = netlink.RouteDel(&delRoute[i]); err != nil {
+			klog.Errorf("delete route item failed, err: %s", err)
```

Comment:
If some error happened, anything else need to be handled besides an error log?

Reply:
route items to delete have no side effects, this statement will be invoked in a new update op(every 100 seconds or Node update).